### DISCUSSION TITLE
Integrate CycloneDX Maven Plugin into build process to generate SBOM?

### DISCUSSION THREAD

**Discussion by hvbtup** [eclipse-birt/birt#1266] [Ideas]
Just for fun I added the following lines into the main `pom.xml` file as described [here](https://github.com/CycloneDX/cyclonedx-maven-plugin#default-values):

			<plugin>
				<groupId>org.cyclonedx</groupId>
				<artifactId>cyclonedx-maven-plugin</artifactId>
				<version>2.7.6</version>
				<executions>
					<execution>
						<phase>package</phase>
						<goals>
							<goal>makeBom</goal>
						</goals>
					</execution>
				</executions>
			</plugin>

This is almost at the end of the file, in the `build` element.

This creates SBOM files in the build folder, eg.

![grafik](https://user-images.githubusercontent.com/7644776/232060664-083d05a2-0177-4acc-907d-96ace8944da4.png)

I am really *not an expert in the areas Maven and SBOM*, but wouldn't that be useful, because clients sometimes demand SBOMs?

Probably the generated SBOM files should be published together with the built artifacts(?).

A drawback of using the plugin is that it slows down the build process.

**Comment by speckyspooky** [comment]
The usage of SBOM like an additional feature makes sense.
Your article is the first time that I heard about SBOM so I studied some articles
and with the middle of last year SBOM is required e.g. on software products which will be used by the US governance.

So it would make sense to do the first steps to be prepared on BIRT side.

Info links:

- SBOM-summary
	- https://www.csoonline.com/de/a/was-ist-eine-software-bill-of-materials,3674040#
- SBOM-products & Best practices (with OWASP CycloneDX)
	- https://www.csoonline.com/de/a/die-8-besten-sbom-tools,3674056
- CyberSecurity & Infrastructure Security Agency (US: CISA): 
	- https://www.cisa.gov/sbom

**Comment by hvbtup** [comment]
Besides, I think that demanding an SBOM (and thus enabling automatic checking for known vulnerabilities) is a much better way to increase software security in comparison to what the EU plans with the Cyber Resiliance Act - which has the potential to throw back Europe 30 years in the field of software, because who is willing to create Open Source software then?

**Comment by wimjongman** [comment]
I like that SBOM feature.

**Comment by hvbtup** [comment]
@waynebeaton This may be an interesting question for other Eclipse projects as well: 
Is there a best practice to generate and publish SBOMs for Eclipse projects?
I saw https://gitlab.eclipse.org/eclipsefdn/emo-team/emo/-/issues/46.

**Reply by waynebeaton** [reply]
At some point, SBOMs are going to become--if not a strict requirement--a strong recommendation for all Eclipse open source projects.

Establishing some best practices is on my list. Right now, I have a lot of observations and opinions, but nothing committed to writing. I did present some of my thoughts on the topic during a recent [Committer Office Hours](https://www.eclipse.org/projects/calendar/#2023-03-07) session. 

One of the challenges is that the Maven plugins that I've looked at only know license and copyright information that they extract from pom.xml files. We've discovered that this information is not always present or accurate. You also have to account for aspects of your build that are funky or non-standard (Maven is super good at identifying all dependencies, but doesn't stand a chance when a build, for example, pulls a JAR apart and repackage parts of it).

The question of where to store them is another challenge. Storing an SBOM in the repository makes it a liability that needs to be maintained (an inaccurate SBOM is worse than no SBOM). Generating it on the fly with a Maven plugin should mitigate against this.

Having said all of that, perfection is the enemy of good, and it's probably better to have something rather than nothing. 

I'll try my best to engage with you on this so that we call all learn some new things.

One think that I am curious about... why CycloneDX rather than SPDX? My experience with CycloneDX is limited; I'm asking only because I'm curious.

**Comment by hvbtup** [comment]
Thanks, @waynebeaton. I just watched the Youtube video you mentioned.
OK, so the EMO is aware of this topic. I'm looking forward for recommendations.

The reason why I chose CycloneDX?

Until now, I only knew SBOM topic from a theoretical point-of-view.

The Log4Shell drama made me spend a whole weekend writing and running ad-hoc tools for scanning the software we delivered to our clients (including examining nested ZIP/JAR/WAR/EAR structures). Fortunately, it turned out that only one of our programs used log4j 1.2 at all, and that the configuration of that particular program was in a way that prevented the security issue. 

So we just had luck then.
Nevertheless, our support team had to answer a lot of questions from our customers for the next 4 weeks.
We replaced the library with a drop-in replacement.

But this incident alerted me: What if next time there is a severe issue in a widely used library version?

We certainly need a way to (at least halfway) *automatically* know which of our programs are using that library version and which of our clients are using those program versions.

Of course this is only a first step, we also need to improve our deployment process to enable us to roll-out new program versions more or less automatically (the software is running on-premise). Fully automatical updates are not possible for many of our clients.
But we don't have the ressources to manually perform several steps for each client (apart from getting access, which takes several minutes each time), so ideally our clients' admins should be able to download and install an update.

At least the first step - knowing about the used libraries in detail - will reduce the possible reaction time from at least 2-3 days to hours.
I think this goal is easy to achieve once we have SBOMs and an internal SBOM repository server.

Being able to generate an SBOM for the 3rdParty libraries, we can probably automatically create SBOMs for our own programs as well.

Our goal with SBOMs is exactly what I described: 
To know which 3rdParty libraries are used by our programs directly and indirectly, including their exact version numbers; and thus reduce reaction time.

I understand that the Eclipse Foundation also needs other information in SBOMS, in particular licensing information.
But for us, this is only nice-to-have.


I only knew there were two "standards" for SBOM: CycloneDX and SPDX.
I spend one day for grepping how this could work in practice.

I had no preference for one or the other format, but I knew that just generating SBOMs is certainly not enough; you also need a repository.

Searching the internet, I found that CycloneDX has software to run such a repository server on-premise.
I couldn't find something similar for SPDX (but it was a brief investigation only).

Most of our software is written in Java or Python.

I started with Python and quickly found a way to generate an CycloneDX SBOM for Python projects.
I couldn't a similar way to generate an SPDX SBOM for Python project.

For Java, I googled a bit and somewhere found a page whic described different possible ways to generate SBOM for Maven projects.
and that CycloneDX works pretty well.

So I started with the CycloneDX Maven plugin. My first attempt nearly killed my disk, because I used an older version of the plugin which ran into some kind of endless directory recursion. I then used the current version and, well, it worked pretty well.

I don't know if it's perfect, maybe some 3rdParty lib is missing. How should I know? I just picked a handful of libraries which I know BIRT is using them and found them all in the SBOM file.

I just did not know that there is also a Maven SPDX plugin.

Do you know if there is a tool which can convert SPDX to CycloneDX and vice versa, keeping the essential information about exact versions intact?

BTW I think that making SBOMs more or less a must-have (as the USA intend to do by an Executive Order) is a much better way to increase software security in comparsion to what the EU tries with the Cyber Resilicance Act. The latter bears a great risk to have a negative effect on security, because in tis currenct form it severely hinders (or prevents) developing and using open source software.
